File: android/app/src/main/java/online/muhta/app/car/CarData.java
package online.muhta.app.car;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationManager;
import android.util.Log;

import androidx.core.content.ContextCompat;

import online.muhta.app.BuildConfig;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * البياناتُ لشاشة السيارة — بلا WebView وبلا Capacitor.
 *
 * <p>شاشةُ السيارة تعمل والتطبيقُ مغلق؛ فلو انتظرت جسرَ Capacitor لَما عملت.
 * فهي تنادي القاعدةَ مباشرةً بنداءٍ واحد: {@code car_nearby}.
 *
 * <p><b>ولا قاعدةَ توفّرٍ هنا.</b> «متوفّرٌ الآن» أربعةُ شروط — متوفّر، وحديثٌ
 * دون أربعٍ وعشرين ساعة، ولم يمرّ موعدُ نفاده، ومحطتُه مفتوحةٌ بتوقيت بغداد —
 * وكلُّها تُحسب في SQL. ولها في المشروع ثلاثُ نسخٍ اليوم وإحداها خاطئة، فنسخةٌ
 * رابعةٌ هنا كانت ستزيد الطينَ بلّة. هذا الملفُّ يجلب ويرسم، ولا يحكم.
 */
final class CarData {

    private static final String TAG = "MuhtaCar";

    /** مركزُ الأنبار — وإليه تُقاس «هل أنت داخل نطاق الخدمة؟». */
    private static final double ANBAR_LAT = 33.4258;
    private static final double ANBAR_LNG = 43.3012;

    /**
     * ثلاثُمئة كيلومترٍ: خارجها لا يُقاس من موقع الجهاز.
     *
     * <p>بياناتُ المنصّة كلُّها داخل محافظةٍ واحدة. فمن يفتح التطبيق في أربيل أو
     * عمّان — أو مراجعُ متجرٍ في كاليفورنيا — قائمتُه فارغةٌ بحقّ، و«فارغة» لا
     * تُقرأ «بعيد» بل تُقرأ «معطّل». فيُقاس من مركز الرمادي، ويُكتب ذلك على
     * الشاشة صراحةً. سلوكٌ صادقٌ لعراقيٍّ مغترب، لا حيلةُ مراجعة.
     */
    private static final double SERVICE_RADIUS_KM = 300;

    /** المهلةُ التي تُعدّ بعدها اللقطةُ قديمةً فتُجلب من جديد. */
    private static final long CACHE_MS = 3 * 60 * 1000L;

    /** صفُّ محطةٍ كما تردّه القاعدة. */
    static final class Station {
        final String id;
        final String name;
        final String city;
        final double lat;
        final double lng;
        final double distanceKm;
        final List<String> products;
        final int confirmedMin;
        /** {@code null} حين تُخفي المحطةُ رقمها ولم تُفتح بوّابةُ المسافر. */
        final String phone;

        Station(String id, String name, String city, double lat, double lng,
                double distanceKm, List<String> products, int confirmedMin, String phone) {
            this.id = id;
            this.name = name;
            this.city = city;
            this.lat = lat;
            this.lng = lng;
            this.distanceKm = distanceKm;
            this.products = products;
            this.confirmedMin = confirmedMin;
            this.phone = phone;
        }
    }

    /** نتيجةُ جلبةٍ واحدة — ومعها سببُ فشلها إن فشلت، وهل قِيست من موقعٍ حقيقيّ. */
    static final class Result {
        final List<Station> stations;
        final boolean outOfRange;
        final String error;

        Result(List<Station> stations, boolean outOfRange, String error) {
            this.stations = stations;
            this.outOfRange = outOfRange;
            this.error = error;
        }
    }

    private static Result cached;
    private static long cachedAt;

    private CarData() {}

    static synchronized void invalidate() {
        cached = null;
        cachedAt = 0;
    }

    /**
     * يُنادى من خيطٍ خلفيّ لا من خيط الواجهة: فيه شبكة.
     */
    static synchronized Result load(Context context) {
        long now = System.currentTimeMillis();
        if (cached != null && now - cachedAt < CACHE_MS) return cached;

        if (BuildConfig.SUPABASE_URL.isEmpty() || BuildConfig.SUPABASE_ANON_KEY.isEmpty()) {
            // إعدادُ البناء ناقص — يُقال ولا يُترك صامتاً.
            return new Result(new ArrayList<>(), false, "الإعداد ناقص — أعِد تثبيت التطبيق");
        }

        Location fix = lastKnown(context);
        boolean outOfRange = true;
        double lat = ANBAR_LAT;
        double lng = ANBAR_LNG;
        if (fix != null) {
            double away = haversineKm(fix.getLatitude(), fix.getLongitude(), ANBAR_LAT, ANBAR_LNG);
            if (away <= SERVICE_RADIUS_KM) {
                lat = fix.getLatitude();
                lng = fix.getLongitude();
                outOfRange = false;
            }
        }

        try {
            Result r = new Result(fetch(lat, lng), outOfRange, null);
            cached = r;
            cachedAt = now;
            return r;
        } catch (Exception e) {
            Log.w(TAG, "car_nearby failed", e);
            // لقطةٌ قديمةٌ خيرٌ من شاشةٍ فارغة — ما دام يُقال إنّها قديمة.
            if (cached != null) return cached;
            return new Result(new ArrayList<>(), outOfRange, "تعذّر الاتصال — تحقّق من الإنترنت");
        }
    }

    /** آخرُ موقعٍ يعرفه النظام. ولا تُطلب أذوناتٌ من شاشة السيارة: تُمنح في الهاتف. */
    private static Location lastKnown(Context context) {
        boolean fine = ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED;
        boolean coarse = ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION)
                == PackageManager.PERMISSION_GRANTED;
        if (!fine && !coarse) return null;

        LocationManager lm = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        if (lm == null) return null;

        Location best = null;
        try {
            for (String provider : lm.getProviders(true)) {
                Location l = lm.getLastKnownLocation(provider);
                if (l == null) continue;
                if (best == null || l.getTime() > best.getTime()) best = l;
            }
        } catch (SecurityException ignored) {
            return null;
        }
        return best;
    }

    private static List<Station> fetch(double lat, double lng) throws Exception {
        URL url = new URL(BuildConfig.SUPABASE_URL + "/rest/v1/rpc/car_nearby");
        HttpURLConnection c = (HttpURLConnection) url.openConnection();
        try {
            c.setRequestMethod("POST");
            c.setConnectTimeout(8000);
            c.setReadTimeout(8000);
            c.setDoOutput(true);
            c.setRequestProperty("Content-Type", "application/json");
            c.setRequestProperty("apikey", BuildConfig.SUPABASE_ANON_KEY);
            c.setRequestProperty("Authorization", "Bearer " + BuildConfig.SUPABASE_ANON_KEY);

            JSONObject body = new JSONObject();
            body.put("p_lat", lat);
            body.put("p_lng", lng);
            body.put("p_limit", 40);
            try (OutputStream os = c.getOutputStream()) {
                os.write(body.toString().getBytes(StandardCharsets.UTF_8));
            }

            int code = c.getResponseCode();
            if (code < 200 || code >= 300) throw new Exception("HTTP " + code);

            StringBuilder sb = new StringBuilder();
            try (BufferedReader r = new BufferedReader(
                    new InputStreamReader(c.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = r.readLine()) != null) sb.append(line);
            }

            JSONArray arr = new JSONArray(sb.toString());
            List<Station> out = new ArrayList<>(arr.length());
            for (int i = 0; i < arr.length(); i++) {
                JSONObject o = arr.getJSONObject(i);
                JSONArray p = o.optJSONArray("products");
                List<String> products = new ArrayList<>();
                if (p != null) for (int j = 0; j < p.length(); j++) products.add(p.getString(j));
                out.add(new Station(
                        o.getString("id"),
                        o.getString("name"),
                        o.optString("city", ""),
                        o.getDouble("lat"),
                        o.getDouble("lng"),
                        o.getDouble("distance_km"),
                        products,
                        o.optInt("confirmed_min", 0),
                        o.isNull("phone") ? null : o.getString("phone")));
            }
            return out;
        } finally {
            c.disconnect();
        }
    }

    static double haversineKm(double aLat, double aLng, double bLat, double bLng) {
        double r = 6371;
        double dLat = Math.toRadians(bLat - aLat);
        double dLng = Math.toRadians(bLng - aLng);
        double h = Math.pow(Math.sin(dLat / 2), 2)
                + Math.cos(Math.toRadians(aLat)) * Math.cos(Math.toRadians(bLat))
                * Math.pow(Math.sin(dLng / 2), 2);
        return 2 * r * Math.asin(Math.sqrt(h));
    }

    /** «١٢٫٣ كم» — رقمٌ عربيٌّ بمنزلةٍ واحدة، كما في بطاقة المحطة على الهاتف. */
    static String km(double v) {
        return String.format(Locale.forLanguageTag("ar"), "%.1f كم", v);
    }

    /**
     * «أُكِّد قبل …» — والصيغةُ تتبع العدد.
     *
     * <p>مرآةُ {@code plural} في {@code lib/freshness.ts}: العربيّةُ تُثنّي
     * وتجمع جمعَ قلّة، و«قبل ٣ ساعة» خطأٌ يقرؤه القارئُ إهمالاً.
     */
    static String since(int minutes) {
        if (minutes < 1) return "أُكِّد الآن";
        if (minutes < 60) return "أُكِّد قبل " + plural(minutes, "دقيقة", "دقيقتين", "دقائق", "دقيقة");
        int h = minutes / 60;
        if (h < 24) return "أُكِّد قبل " + plural(h, "ساعة", "ساعتين", "ساعات", "ساعة");
        int d = h / 24;
        return "أُكِّد قبل " + plural(d, "يوم", "يومين", "أيام", "يوماً");
    }

    private static String plural(int n, String one, String two, String few, String many) {
        if (n == 1) return one;
        if (n == 2) return two;
        int mod = n % 100;
        if (mod >= 3 && mod <= 10) return n + " " + few;
        return n + " " + many;
    }
}

File: ios/App/App/CarData.swift
import CoreLocation
import Foundation

/// البياناتُ لشاشة CarPlay — بلا WebView وبلا Capacitor.
///
/// شاشةُ السيارة تعمل والتطبيقُ مغلق، فلو انتظرت جسرَ Capacitor لَما عملت.
/// فهي تنادي القاعدةَ مباشرةً بنداءٍ واحد: `car_nearby`.
///
/// **ولا قاعدةَ توفّرٍ هنا.** «متوفّرٌ الآن» أربعةُ شروط — متوفّر، وحديثٌ دون
/// أربعٍ وعشرين ساعة، ولم يمرّ موعدُ نفاده، ومحطتُه مفتوحةٌ بتوقيت بغداد —
/// وكلُّها تُحسب في SQL. ولها في المشروع ثلاثُ نسخٍ اليوم وإحداها خاطئة، فنسخةٌ
/// رابعةٌ هنا كانت ستزيد الطينَ بلّة. هذا الملفُّ يجلب ويرسم، ولا يحكم.
enum CarData {

    /// مركزُ الأنبار — وإليه تُقاس «هل أنت داخل نطاق الخدمة؟».
    static let anbar = CLLocationCoordinate2D(latitude: 33.4258, longitude: 43.3012)

    /// ثلاثُمئة كيلومترٍ: خارجها لا يُقاس من موقع الجهاز.
    ///
    /// بياناتُ المنصّة كلُّها داخل محافظةٍ واحدة. فمن يفتح التطبيق في أربيل أو
    /// عمّان — أو مراجعُ متجرٍ في كاليفورنيا — قائمتُه فارغةٌ بحقّ، و«فارغة» لا
    /// تُقرأ «بعيد» بل تُقرأ «معطّل». فيُقاس من مركز الرمادي، ويُكتب ذلك على
    /// الشاشة صراحةً. سلوكٌ صادقٌ لعراقيٍّ مغترب، لا حيلةُ مراجعة.
    static let serviceRadiusKm: Double = 300

    /// المهلةُ التي تُعدّ بعدها اللقطةُ قديمةً فتُجلب من جديد.
    static let cacheSeconds: TimeInterval = 180

    struct Station {
        let id: String
        let name: String
        let city: String
        let lat: Double
        let lng: Double
        let distanceKm: Double
        let products: [String]
        let confirmedMin: Int
        /// `nil` حين تُخفي المحطةُ رقمها ولم تُفتح بوّابةُ المسافر.
        let phone: String?
    }

    struct Result {
        let stations: [Station]
        let outOfRange: Bool
        let error: String?
    }

    private static var cached: Result?
    private static var cachedAt: Date?
    private static let manager = CLLocationManager()

    static func invalidate() {
        cached = nil
        cachedAt = nil
    }

    /// العنوانُ والمفتاحُ العامّ من `Info.plist` — يُحقنان عند البناء ولا
    /// يُكتبان في المصدر. والمستودعُ عامّ، وقد كلّف ذلك الدرسُ ثمنَه مرّةً.
    private static func setting(_ key: String) -> String {
        (Bundle.main.object(forInfoDictionaryKey: key) as? String) ?? ""
    }

    static func load(_ done: @escaping (Result) -> Void) {
        if let c = cached, let at = cachedAt, Date().timeIntervalSince(at) < cacheSeconds {
            done(c)
            return
        }

        let base = setting("MuhtaSupabaseURL")
        let key = setting("MuhtaSupabaseAnonKey")
        guard !base.isEmpty, !key.isEmpty,
              let url = URL(string: base + "/rest/v1/rpc/car_nearby") else {
            done(Result(stations: [], outOfRange: false, error: "الإعداد ناقص — أعِد تثبيت التطبيق"))
            return
        }

        // لا نافذةَ إذنٍ من شاشة السيارة: CarPlay لا يستطيع عرضَها، والإذنُ
        // يُمنح في الهاتف مرّةً. فيُقرأ ما هو ممنوحٌ الآن، ولا يُطلب شيء.
        var lat = anbar.latitude
        var lng = anbar.longitude
        var outOfRange = true
        // `manager.authorizationStatus` لا `CLLocationManager.authorizationStatus()`:
        // الثانيةُ مهجورةٌ منذ iOS 14 وقد تُرفض على SDK حديث، والهدفُ 15.
        let status = manager.authorizationStatus
        if status == .authorizedWhenInUse || status == .authorizedAlways,
           let here = manager.location {
            let away = haversineKm(here.coordinate.latitude, here.coordinate.longitude,
                                   anbar.latitude, anbar.longitude)
            if away <= serviceRadiusKm {
                lat = here.coordinate.latitude
                lng = here.coordinate.longitude
                outOfRange = false
            }
        }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.timeoutInterval = 8
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(key, forHTTPHeaderField: "apikey")
        req.setValue("Bearer " + key, forHTTPHeaderField: "Authorization")
        req.httpBody = try? JSONSerialization.data(withJSONObject: [
            "p_lat": lat, "p_lng": lng, "p_limit": 40,
        ])

        URLSession.shared.dataTask(with: req) { data, response, _ in
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard code >= 200, code < 300, let data = data,
                  let raw = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
                // لقطةٌ قديمةٌ خيرٌ من شاشةٍ فارغة — ما دام يُقال إنّها قديمة.
                DispatchQueue.main.async {
                    done(cached ?? Result(stations: [], outOfRange: outOfRange,
                                          error: "تعذّر الاتصال — تحقّق من الإنترنت"))
                }
                return
            }

            let stations: [Station] = raw.compactMap { o in
                // **والإحداثيّةُ لا تُستبدل بصفر.** (0,0) نقطةٌ صالحةٌ في المحيط
                // الأطلسيّ قبالة غانا — فصفٌّ ناقصٌ كان يصير محطةً على بُعد
                // خمسة آلاف كيلومتر، ويُسلَّم دليلاً إلى خرائط آبل. فيُسقَط.
                guard let id = o["id"] as? String, let name = o["name"] as? String,
                      let lat = o["lat"] as? Double, let lng = o["lng"] as? Double else { return nil }
                return Station(
                    id: id,
                    name: name,
                    city: o["city"] as? String ?? "",
                    lat: lat,
                    lng: lng,
                    distanceKm: o["distance_km"] as? Double ?? 0,
                    products: o["products"] as? [String] ?? [],
                    confirmedMin: o["confirmed_min"] as? Int ?? 0,
                    phone: o["phone"] as? String)
            }

            let result = Result(stations: stations, outOfRange: outOfRange, error: nil)
            cached = result
            cachedAt = Date()
            DispatchQueue.main.async { done(result) }
        }.resume()
    }

    static func haversineKm(_ aLat: Double, _ aLng: Double, _ bLat: Double, _ bLng: Double) -> Double {
        let r = 6371.0
        let dLat = (bLat - aLat) * .pi / 180
        let dLng = (bLng - aLng) * .pi / 180
        let h = pow(sin(dLat / 2), 2)
            + cos(aLat * .pi / 180) * cos(bLat * .pi / 180) * pow(sin(dLng / 2), 2)
        return 2 * r * asin(sqrt(h))
    }

    /// «١٢٫٣ كم» بمنزلةٍ واحدة، كما في بطاقة المحطة على الهاتف.
    static func km(_ v: Double) -> String {
        String(format: "%.1f كم", v)
    }

    /// «أُكِّد قبل …» — والصيغةُ تتبع العدد.
    ///
    /// مرآةُ `plural` في `lib/freshness.ts`: العربيّةُ تُثنّي وتجمع جمعَ قلّة،
    /// و«قبل ٣ ساعة» خطأٌ يقرؤه القارئُ إهمالاً.
    static func since(_ minutes: Int) -> String {
        if minutes < 1 { return "أُكِّد الآن" }
        if minutes < 60 { return "أُكِّد قبل " + plural(minutes, "دقيقة", "دقيقتين", "دقائق", "دقيقة") }
        let h = minutes / 60
        if h < 24 { return "أُكِّد قبل " + plural(h, "ساعة", "ساعتين", "ساعات", "ساعة") }
        return "أُكِّد قبل " + plural(h / 24, "يوم", "يومين", "أيام", "يوماً")
    }

    private static func plural(_ n: Int, _ one: String, _ two: String,
                               _ few: String, _ many: String) -> String {
        if n == 1 { return one }
        if n == 2 { return two }
        let mod = n % 100
        return "\(n) " + ((mod >= 3 && mod <= 10) ? few : many)
    }
}

/// أسماءُ الوقود وترتيبُه — مرآةُ `PRODUCT_ORDER` و`PRODUCT_LABELS` في
/// `lib/products.ts`. ولا سبيلَ إلى استيراد TypeScript في Swift، فتُقال صراحةً.
///
/// **و«كاز» يبقى.** وقعت مشورةٌ بحذفه بحجّة أنّه وقودُ تدفئةٍ لا وقودُ سيّارة،
/// والقياسُ يقول غيرَ ذلك: يومَ كُتب هذا كانت إحدى عشرةَ محطةَ كازٍ مؤكَّدةً
/// حديثاً مقابل ثلاثٍ للبانزين العاديّ. فحذفُه كان سيُفرغ الشاشة.
enum CarProducts {
    static let order = [
        "gasoline_regular", "gasoline_premium", "gasoline_super",
        "kerosene", "gas", "lpg", "white_oil",
    ]

    static let labels: [String: String] = [
        "gasoline_regular": "بانزين عادي",
        "gasoline_premium": "بانزين محسن",
        "gasoline_super": "بانزين سوبر",
        "kerosene": "كاز",
        "gas": "غاز",
        "lpg": "LPG",
        "white_oil": "نفط أبيض",
    ]

    static func label(_ key: String) -> String { labels[key] ?? key }

    static func join(_ keys: [String]) -> String {
        order.filter { keys.contains($0) }.map(label).joined(separator: " · ")
    }
}
